### PR TITLE
Sanitize HTML for documentation from translators markdown

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
Follow up to #15945 

### Summary of the issue:
Translators can add arbitrary HTML to markdown translations files.
This is a [stored XSS risk](https://owasp.org/www-community/attacks/xss/)

### Description of user facing changes
Should be none - however see known issues

### Description of development approach
Used the [python bindings](https://nh3.readthedocs.io) for the [Rust Ammonia sanitization library](https://docs.rs/ammonia/)

### Testing strategy:
Tested building docs, diffed HTML results between this PR and beta.

### Known issues with pull request:
The sanitization deletes any HTML tags that are not recognized.
This includes using angle brackets around words, e.g.: ``<minor>``.
If these are wrapped by code formatting these are correctly escaped: e.g. ``` `<minor>` ```
While english files are mostly correct, many translated files do not wrap text with angle brackets with code formatting.
This means certain parts of translated documentation will be stripped, i.e. `<major>.<minor>.<patch>` becomes `..`

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -24,6 +24,8 @@ Markdown==3.5.1
 mdx_truly_sane_lists==1.3
 markdown-link-attr-modifier==0.2.1
 mdx-gh-links==0.4
+# Sanitize HTML documention output to prevent XSS from translators
```
Comment (michaelDCurran) [inline requirements.txt:27]:
```suggestion
# Sanitize HTML documentation output to prevent XSS from translators
```

**Review by michaelDCurran — DISMISSED** [review]

**Comment by CyrilleB79** [comment]
@seanbudd, you write:
> The sanitization deletes any HTML tags that are not recognized. This includes using angle brackets around words, e.g.: `<minor>`. If these are wrapped by code formatting these are correctly escaped: e.g. `` `<minor>` `` While english files are mostly correct, many translated files do not wrap text with angle brackets with code formatting. This means certain parts of translated documentation will be stripped, i.e. `<major>.<minor>.<patch>` becomes `..`

I can see that you have fixed such issues in the Russian change log. But I guess other languages are impacted.
Is there an easy way for translators to figure which part of the file has been modified/deleted so that they can fix it? Else, we will probably end up with many information removed from the translated documents.

Though, I understand the need to sanitize documentation's HTML; my remark is not a criticism of this approach.

**Comment by seanbudd** [comment]
>Is there an easy way for translators to figure which part of the file has been modified/deleted so that they can fix it?

Searching for angle brackets in markdown files should make it easy to find these issues.
Some early bad news: Translators will be needing to re-check the entire translated documents line by line when these files go up in crowdin. This means every line will need to be manually checked anyway.

The iframe tag in the Russian source caused the rest of the document to be escaped - i.e. half the document appeared as HTML source code in the browser. For other instances, it is just the fake "tag" that was removed.

File: site_scons/site_tools/md2html.py [hunk 2]
```diff
@@ -68,75 +81,108 @@ def _getTitle(mdBuffer: io.BytesIO, isKeyCommands: bool = False) -> str:
...
 		# Remove heading hashes and trailing whitespace to get the tab title
-		title = mdBuffer.readline().decode("utf-8").strip().lstrip("# ")
+		title = mdBuffer.readline().strip().lstrip("# ")
 
 	return title
 
 
+def _createAttributeFilter() -> dict[str, set[str]]:
+	# Create attribute filter exceptions for HTML sanitization
+	import nh3
```
Comment (gerald-hartig) [inline site_scons/site_tools/md2html.py:91]: Should this import go at the top? PEP 8 advises that imports should usually be placed at the top of a file, except in certain cases like circular import issues and optimising performance when using large libraries that are not always used.

File: site_scons/site_tools/md2html.py [hunk 3]
```diff
@@ -68,75 +81,108 @@ def _getTitle(mdBuffer: io.BytesIO, isKeyCommands: bool = False) -> str:
...
+
+	return allowedAttributes
+
+
+ALLOWED_ATTRIBUTES = _createAttributeFilter()
+
+
+def _generateSanitizedHTML(md: str, isKeyCommands: bool = False) -> str:
+	import markdown
+	import nh3
```
Comment (gerald-hartig) [inline site_scons/site_tools/md2html.py:120]: As above

File: site_scons/site_tools/md2html.py (hunk 2 above)
Comment (seanbudd) [reply site_scons/site_tools/md2html.py:91]: This can't be done in these scons scripts

**Pull request merged**